### PR TITLE
[WIP] Support Eigen::Map<Eigen::SparseMatrix>

### LINKED ISSUE #781: [BUG]: scipy.sparse._csc.csc_matrix → Eigen::SparseMatrix appears to incur a copy
### Problem description

The [docs](https://nanobind.readthedocs.io/en/latest/eigen.html#sparse-matrices) write that Eigen::SparseMatrix _maps_ to scipy.sparse.csc/csr, but in my minimal example seems to show that it _copies_. I hope I'm mistaken and this is not actually a bug, in that case apologies I didn't make this a discussion question.

[pybind11's documentation](https://pybind11.readthedocs.io/en/stable/advanced/cast/eigen.html#pass-by-reference) has a note that seems to imply that pass by reference is not supported for Eigen::SparseMatrix so I wonder if #126 inherited that.

If there's no current intention to support passing SparseMatrices efficiently, perhaps this is simply a documentation bug. In that case, I wonder if I could help try to add support. My understanding was that [numpyeigen](https://github.com/fwilliams/numpyeigen/blob/master/src/npe_sparse_array.h) passed without a copy (I will try to verify).

### Reproducible example code

```text
#include <nanobind/nanobind.h>
#include <nanobind/ndarray.h>
#include <nanobind/eigen/dense.h>
#include <nanobind/eigen/sparse.h>

namespace nb = nanobind;

nb::object foo(const Eigen::SparseMatrix<double>& X)
{
  return nb::none();
}

// Bind the wrapper to the Python module
NB_MODULE(my_module, m) 
{
  m.def("foo", &foo, nb::arg("X"));
}


-------------------------------
import numpy as np
import my_module
import time
import scipy.sparse
# timing
import time

def rand_sparse(n,density):
    n_features = n
    n_samples = n
    rng1 = np.random.RandomState(42)
    rng2 = np.random.RandomState(43)

    nnz = int(n_samples*n_features*density)

    row = rng1.randint(n_samples, size=nnz)
    cols = rng2.randint(n_features, size=nnz)
    data = rng1.rand(nnz)

    S = scipy.sparse.csc_matrix((data, (row, cols)), shape=(n_samples, n_features))
    return S

for p in range(2,8):
    n = 10**p
    S = rand_sparse(n,1.0/(n))
    my_module.foo(S)
    start = time.time()
    for i in range(10):
        my_module.foo(S)
    end = time.time()
    t = (end - start)/10
    print("%10d %f" % (n,t))
```

### PR DESCRIPTION
**Pull request by alecjacobson**
If completed would fix #781.

Phew, I learned a lot poking around, but I'm mostly just cribbing the original Eigen::SparseMatrix support. I have barely tested any of this. I will use this PR to add tests etc.

### CODE REVIEW

**Comment by wjakob** [comment]
Hi Alec, thank you very much for working on this, zero-copy sparse matrix passing will be useful. There were still some internal comments and questions inside the PR, so I am assuming that it is still in progress. Please let me know if/when you'd like me to do a review.

File: include/nanobind/eigen/sparse.h [hunk 1]
```diff
@@ -146,15 +146,116 @@ template <typename T> struct type_caster<T, enable_if_t<is_eigen_sparse_matrix_v
...
+    Index rows, cols, nnz;
+
+    bool from_python(handle src, uint8_t flags, cleanup_list *cleanup) noexcept 
+    {
+        object obj = borrow(src);
+        try {
+            object matrix_type = module_::import_("scipy.sparse").attr(RowMajor ? "csr_matrix" : "csc_matrix");
+            if (!obj.type().is(matrix_type))
+                obj = matrix_type(obj);
+        } catch (const python_error &) {
```
[the remark this answers is not in the document]
Comment (bilmes) [reply include/nanobind/eigen/sparse.h:189]: Sorry, another minor comment on the above. If it is the case that you do end up returning false if `!obj.type().is(matrix_type)` then the catch block of the exception I'd guess should probably re-throw the exception since it probably means something is wrong other than that the object is not a sparse matrix (e.g., an import or some other error). I.e., the return of true/false should probably (unless I don't understand things) only happen if there are no errors.  Similarly, in the below, when you are returning false, this means that it passes the above type check but there is still a problem getting, say, indices, and I'd guess that would mean something is quite wrong (not well indicated by a false return) with the object and/or setup.

**Comment by alecjacobson** [comment]
I got really close then ran out of time to finish. I don't think there's
much left before asking for a review.

File: include/nanobind/eigen/sparse.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (wjakob) [reply include/nanobind/eigen/sparse.h:189]: A point from me would be that the `Map` casters are designed to really map the data type from one framework into the other without any kind of conversion/copy. It's the user's job to ensure that the data is compatible, since implicit conversions are turned off. For the dense caster, that means that strides and data type are correct. For the sparse caster, it therefore also seems to me that we should already be in the right sparse format so that no copy is needed.

**Comment by wjakob** [comment]
Hi Alec,

I took a look at this PR today and left a comment above. Regarding these two comments:

```cpp
// Disable implicit conversions
//
// I'm not convinced this manipulation of the flags works. It seems in
// ndarray caster the flag is just reset to true.
```

and

```cpp
// I thought this would not allow conversions, but it does
// I think conversion is ok if std::is_const_v<T> is true, so long as
// each *_caster is the owner.
// For non-const, conversion seems to be a bad idea. I think the dense
// version will throw a RunTime error rather than convert
```

I am wondering what made you think that? I looked through the ndarray caster
and don't find this "reset to true" part. In the meantime, there have been
various fixes to both ndarray and Eigen type casters, so it's also possible
that this is a comment about an older version that is moot once you rebase
(which should be easy, there are no conflicts).

Did the libIGL port to nanobind go OK? I saw the huge PR https://github.com/libigl/libigl-python-bindings/pull/243.

**Comment by alecjacobson** [comment]
Thanks so much for looking at this Wenzel. I hope to make time to complete it but it's unclear how soon that would be. If anyone else reading this wants to jump in on finishing or collaborating on finishing it, I'd be happy/grateful. 

The libigl port to nanobind went well and very soon will be the official release (it's been the dev release on pypi for some months). I removed a lot of really messy stuff on the old pybind version of the libigl bindings. The sparse matrices are copied (relying on this PR), but otherwise the bindings are pretty clean. I think the only thing I'm not in love with is the automatic generation of adding the bindings to the module, but I'm sure this is all my fault not nanobinds.

**Comment by wjakob** [comment]
Hi Alec,

That's great to hear! Two quick questions though. First, I was wondering what this means?

> automatic generation of adding the bindings to the module

Are you using [litgen](https://pthom.github.io/litgen/litgen_book/00_00_intro.html) or some tool like that?

And a follow up regarding this:

> If anyone else reading this wants to jump in on finishing or collaborating on finishing it, I'd be happy/grateful.

If it's just a little bit of work left, I could also take a look. But for that, a quick pass from you would be needed:

- Can you take a look at the discussion here and write a quick response to questions that were brought up?
- From your memory, what is still missing? What did not work?
- Are there important tests that are missing?

Basically a braindump so that someone else (e.g. me) can pick up, otherwise a big part of the time will just be figuring out what to do.

File: include/nanobind/eigen/sparse.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (alecjacobson) [reply include/nanobind/eigen/sparse.h:189]: I haven't tested @bilmes fix, but I agree with this.

**Comment by alecjacobson** [comment]
re:libigl bindings. Basically, for each libigl function there's a bit of boilerplate I need to do before applying nanobind (mostly to translate output-via-parameter to output-via-return, but maybe like 20% of the functions also required some other bespoke handling). So, to organize all the bindings I put them each in a corresponding .cpp file. See https://github.com/libigl/libigl-python-bindings/blob/main/src/barycenter.cpp for example. That file defines `bind_barycenter` which takes a module reference as arg. Then during cmake all these `bind_*` are gathered to populate into the `NB_MODULE` call in `https://github.com/libigl/libigl-python-bindings/blob/main/src/module.cpp`. I find this kind of gross but acceptable (I needed to move on to actually rewriting all the bindings). At the very least, I like that adding a new binding means just adding a new file like https://github.com/libigl/libigl-python-bindings/blob/main/src/barycenter.cpp which doesn't have much "magic" looking stuff in it.

re:this pr
Right, I replied about the conversion. I agree this shouldn't be trying to copy/convert dense to sparse. I think my intention at the time was to allow CSC<->CSR to happen via a copy, but it makes sense that this is not the point of Map. From my memory, the main thing missing was feeling confident that the set of tests ensured that the correct set of things are/aren't copying. I doubt datatypes will be a problem but probably at least a std::complex would be good to add. 

I'm sorry I don't 100% remember but I do remember being convinced that `ndarray caster the flag is just reset to true` happened in the code paths I followed. If I got time to dig this out I could try to convince myself, but maybe it's moot now.

**Comment by wjakob** [comment]
> I find this kind of gross but acceptable (I needed to move on to actually rewriting all the bindings). At the very least, I like that adding a new binding means just adding a new file like https://github.com/libigl/libigl-python-bindings/blob/main/src/barycenter.cpp which doesn't have much "magic" looking stuff in it.

I took a look and it's at least easy to understand. If the replication to write another `pyigl::barycenter` bothers you, you could probably make a template that constructs this version for you automatically. It would basically do a pass over the function arguments and replace each one by `nb::DRef<const Eigen::MatrixXN>` etc. as needed. The main difficulty here that I see is that your _original_ C++ libigl template declarations don't specify enough context to be able to fully automate such rewriting. It's not clear from the function definition which arguments are integers vs floating point arrays, how many dimensions, etc. So to handle rewriting in this generality, you would need several rewriting templates to handle different patterns, at which point what you are doing right now does not look so bad. So this is more of a suggestion for the next major libigl refactor..

**Comment by wjakob** [comment]
By the way, some people have reported great success in using LLVM's to write nanobind bindings. Just copy-paste your original code and say "please nanobind it!" ;-). With extra libigl-specific prompting, it could likely largely eliminate you having to manually write this code.
